### ISSUE TITLE
hasId should be optimized to perform in O(1)

### ISSUE THREAD

**Issue by porunov** [JanusGraph/janusgraph#1525]
Currently `hasId` step performs in O(n) complexity but should be O(1).
Currently `is` step complexity is O(1) but `hasId` step is O(n) which is strange.
I.e. with the big amount of outgoing edges the next code will take a lot of time:
```
Vertex outVertex = ...;
Vertex inVertex = ...;
graph.traversal().V(outVertex.id()).outE("my_label").where(__.otherV().hasId(inVertex.id())).hasNext();
```
But the same query with `is` will perform almost immediately: 
```
Vertex outVertex = ...;
Vertex inVertex = ...;
graph.traversal().V(outVertex.id()).outE("my_label").where(__.otherV().is(inVertex)).hasNext();
```

**Comment by rngcntr** [comment]
I'll have a closer look at this as it's related to #1937 

At this point, I don't think this is a question of complexity. From what I can tell, both queries are O(n), but one of them uses the information encoded in the edges while the other one fetches the adjacent vertex of every edge. This should be easily avoidable.

**Comment by porunov** [comment]
Fixed by #1960
